text/1005-edgedb-project.rst

```diff
@@ -0,0 +1,418 @@
+RFC 1005: CLI and conventions for local projects using EdgeDB
+=============================================================
+
+This RFC describes the design of the ``edgedb project`` CLI commands for
+the purposes of initialization of a software project that uses EdgeDB as
+well as the associated conventions and changes to the language bindings.
+
+
+Motivation
+==========
+
+`RFC 1001 <1001-edgedb-server-control.rst>`_ introduced the concept of
+EdgedDB *instances* and a straightforward way of managing them with
+edgedb-cli.  The process of addressing an EdgeDB server has thus been
+simplified to passing a ``-I`` argument to the CLI or passing it as the
+argument to a ``connect()`` call in bindings.
+
+However, even with ``edgedb server`` a developer needs to take multiple
+steps to start a new project using EdgeDB:
+
+1. Install the desired EdgeDB version (``edgedb server install``)
+2. Initialize an EdgeDB instance for the project (``edgedb server init myapp``)
+3. Create the ``dbschema/`` directory and the initial schema file.
+
+When starting work on a cloned project, the steps are similar:
+
+1. Install the desired EdgeDB version (``edgedb server install``)
+2. Initialize an EdgeDB instance for the project (``edgedb server init myapp``)
+3. Apply migrations (``edgedb -I myapp migrate``).
+
+Since most projects using EdgeDB are expected to follow the conventions
+established by the reasonable defaults of the EdgeDB CLI it makes sense to
+automate the above operations to enable the following benefits:
+
+1. Make it so that the project is ready to go after a single command
+   (``edgedb project init``).
+2. Allow an EdgeDB instance to be linked with a project directory to obviate
+   the need to specify an instance name explicitly when running CLI commands
+   or connecting with language bindings.
+3. Allow specifying certain EdgeDB-specific metadata per project, such as
+   the minimum required EdgeDB version.
+
+
+Overview
+========
+
+In this RFC we propose a new group of ``edgedb`` CLI commands under
+the ``edgedb project`` prefix:
+
+* ``edgedb project init`` -- populate a new project or initialize an existing
+  cloned object;
+
+* ``edgedb project deinit`` -- remove association with and optionally destroy
+  the linked EdgeDB instance;
+
+* ``edgedb project status`` -- show EdgeDB-related information about a
+  project, such as the associated instance name;
+
+* ``edgedb project list`` -- list all known EdgeDB projects for the current
+  user.
+
+
+Design
+======
+
+Project file structure
+----------------------
+
+`RFC 1000 <1000-migrations.rst>`_ defined the top-level ``dbschema/`` directory
+as the default location of EdgeDB SDL schema, and ``dbschema/migrations`` as
+the default expected location of the files containing schema migration scripts.
+To mark a directory as an EdgeDB project, an ``edgedb.toml`` file should be
+present in the root directory.  The file can be empty or contain the following
+TOML::
+
+    [edgedb]
```

Comment:
What does the section means? Is this sorta comment? Does it mean we should put everything under `edgedb` in the future, i.e. `[edgedb.extensions]`, etc.

Reply:
We could call this `[server]` for clarity.

> Does it mean we should put everything under `edgedb` in the future, i.e. `[edgedb.extensions]`, etc.

If it's server-related, sure.  We can potentially have other sections that drive the CLI or other clients.